comprobo_final_project/scripts/gene_alg_2/generation.py

```diff
@@ -0,0 +1,157 @@
+
+                # Random chance to mutate either child
+                for child in children:
+                    if np.random.rand() <= self.mutation_thresh:
+                        buf.append(self._mutate(child))
+                    else:
+                        buf.append(self._mutate(child))
+                num_organisms_created += 2
+
+            # Directly move a past organism to the next generation with
+            # a chance at mutation
+            else:
+                curr_organism = self._organisms[num_organisms_created]
+                if np.random.rand() <= self.mutation_thresh:
+                    buf.append(self._mutate(curr_organism))
+                else:
+                    buf.append(curr_organism)
+                num_organisms_created += 1
+
+        # Sort current generation by fitness
+        self._organisms = np.asarray(buf)
+
+
+    def _sort(self):
+        """
+        Sorts organisms by fitness.
+        """
+        order = self._fitnesses.argsort()
+        self._organisms = self._organisms[order]
+        self._fitnesses = self._fitnesses[order]
+
+
+    def _tournament_selection(self):
+        """
+        A helper method used to select a random organism from the
+        generation using a tournament selection algorithm.
+        """
+        choices = np.random.choice(self.gen_size, 5)
+        choice = np.argmin(self._fitnesses[choices])
+        return self._organisms[choices][choice]
+
+
+    def _select_parents(self):
+        """
+        A helper method used to select two parents from the generation using a
+        tournament selection algorithm.
+        """
+        return (self._tournament_selection(), self._tournament_selection())
+
+
+    def _crossover(self, organism_1, organism_2):
+        """
+        Mixes the two specified organisms, returning two new organisms
+        that are a result of a crossover of the two original organisms.
+
+        other: second organism to crossover
+
+        return: two organisms that are crossovers between self and other
+        """
+
+        org_1 = organism_1[:]
+        org_2 = organism_2[:]
+
+        # Define a random pivot point around which the crossover will occur
+        pivot = np.random.randint(0, self.num_genes-1)
+
+        # Create the new crossovered genes and organism
+        right_genes = org_2[:pivot].copy()
+        org_2[:pivot], org_1[:pivot] = org_1[:pivot], right_genes
+
+        return org_1, org_2
+
+
+    def _mutate(self, organism):
+        """
+        Mutates a single random gene of the specified organism.
+        """
+
+        # Don't modify existing organism
+        _organism = organism[:]
+
+        # Select a random gene and multiply it with a random value
+        index_to_mutate = np.random.randint(0, len(_organism) - 1)
+        _organism[index_to_mutate] *= np.random.uniform(0.5, 2)
+
+        # Clip and round all genes
+        _organism[index_to_mutate] = np.clip(_organism[index_to_mutate],
```

Comment:
Yes? I didn't really write that code, also it would only go over during a mutate, so I would assume so.